src/lib/shmem/src/allocator.rs

```diff
@@ -0,0 +1,252 @@
+    /// via other [`ShMemBlock`] objects that alias this one.
+    pub unsafe fn deallocate(mut self) {
+        match self.allocator_or_serializer {
+            AllocatorOrSerializer::Allocator(a) => {
+                unsafe {
+                    c_bindings::shmemallocator_free(a.internal, &mut self.internal as *mut _)
+                };
+            }
+            AllocatorOrSerializer::Serializer(_) => {
+                panic!("Cannot deallocate a block that was instantiated via a Serializer");
+            }
+        }
+    }
+
+    pub fn serialize(&self) -> ShMemBlockSerialized {
+        let serialized = match self.allocator_or_serializer {
+            AllocatorOrSerializer::Allocator(a) => unsafe {
+                c_bindings::shmemallocator_blockSerialize(a.internal, &self.internal)
+            },
+            AllocatorOrSerializer::Serializer(s) => unsafe {
+                c_bindings::shmemserializer_blockSerialize(s.internal, &self.internal)
+            },
+        };
+        ShMemBlockSerialized {
+            internal: serialized,
+        }
+    }
+
+    // This function will fail to compile for a T that isn't FFI-safe.
+    #[deny(improper_ctypes_definitions)]
+    extern "C" fn _validate_stable_layout(_: T) {}
+}
+
+impl<'a, T> Deref for ShMemBlock<'a, T>
+where
+    T: Sync + VirtualAddressSpaceIndependent,
+{
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        unsafe { (self.internal.p as *const T).as_ref().unwrap() }
+    }
+}
+
+pub struct ShMemBlockSerialized {
+    internal: c_bindings::ShMemBlockSerialized,
+}
+
+impl ShMemBlockSerialized {
+    // Keep in sync with macro of same name in shmem_allocator.h.
+    const SHD_SHMEM_BLOCK_SERIALIZED_MAX_STRLEN: usize = 21 + 21 + 21 + 256 + 1;
+
+    pub fn to_string(&self) -> String {
+        let mut buf = Vec::new();
+        buf.resize(Self::SHD_SHMEM_BLOCK_SERIALIZED_MAX_STRLEN, 0i8);
+        unsafe { c_bindings::shmemblockserialized_toString(&self.internal, buf.as_mut_ptr()) };
+        let buf = buf
+            .iter()
+            .take_while(|c| **c != 0)
+            .map(|c| *c as u8)
+            .collect();
+        String::from_utf8(buf).unwrap()
+    }
+
+    pub fn from_string(s: &str) -> Option<Self> {
+        let mut err: bool = false;
+        let mut buf: Vec<i8> = s.as_bytes().iter().map(|b| *b as i8).collect();
+        // Null terminate.
+        buf.push(0);
+        let res = Self {
+            internal: unsafe {
+                c_bindings::shmemblockserialized_fromString(buf.as_ptr(), &mut err)
+            },
+        };
+        if err {
+            None
+        } else {
+            Some(res)
+        }
+    }
+}
+
+pub struct Allocator {
+    internal: *mut c_bindings::ShMemAllocator,
+}
+
+// SAFETY: The C bindings for ShMemAllocator use internal locking to ensure
+// thread-safe access.
+unsafe impl Send for Allocator {}
+unsafe impl Sync for Allocator {}
```

Comment:
optional nit: Maybe put `*mut c_bindings::ShMemAllocator` into a `SyncSendPointer` instead to make it more explicit that the only unsafe part is making the pointer `Send` and `Sync`. (And if anything else that's not `Send` and `Sync` was added to `Allocator` in the future, it would correctly cause `Allocator` to become not `Send + Sync`.)

Reply:
Yeah, that'd be better, but would require reorganizing crates to avoid a circular dependency. I put a TODO for now.